ejb3/src/main/java/org/jboss/as/ejb3/remote/EJBRemotingConnectorClientMappingsEntryProviderService.java

```diff
@@ -101,7 +102,11 @@ public ServiceConfigurator configure(OperationContext context) {
             // This needs to be configurable (i.e. send either host name or the IP address). But
             // since this is a corner case (i.e. absence of any client-mappings for a socket binding),
             // this should be OK for now
-            final String destinationAddress = info.getSocketBinding().getAddress().getHostAddress();
+            InetAddress addr = info.getSocketBinding().getAddress();
+            String destinationAddress = addr.getHostAddress();
+            if (addr.isAnyLocalAddress() && !this.group.get().isSingleton()) {
```

Comment:
I don't think this is the best place for this validation logic.  Surely any remote connector should not accept a socket binding that uses a wildcard address.  So, perhaps this validation logic should be moved to the remoting subsystem.

Reply:
I don't see why it is invalid for a remote connector to bind to the wildcard address. The final connection dictates the true topology to which the logical EJB connection must abide.

Reply:
Two reasons:
* The RemotingConnectorBindingInfoService.RemotingConnectorInfo does not expose that information.
* Client mappings are specific to the specific interface on which the given remote connector is bound.

Reply:
I think there is no cluster information in the remoting subsystem(because it is in wildfly-core), so I assume you are suggesting checking only on the wildcard address binding no matter HA profile is used or not. If that is the case, the checking will make the command `bin/standalone.sh -b 0.0.0.0` fail except we specify how to choose an address for the remote connector to bind to.

> Surely any remote connector should not accept a socket binding that uses a wildcard address. 

`http-remoting-connector` shares the same `SocketBinding` as the HTTP one, disable wildcard address binding to `http-remoting-connector` will also disable such binding to HTTP.

Reply:
Thinking about it a bit more, does a serialized EJB proxy contain an `URL`?
It should have one to reconnect to the actual EJB (regardless of whether it is one running on a node or cluster).
I can agree that having `INADDR_ANY` in that `URL` makes no sense, so that would be a bug. It should be something else, potentially a DNS name or a logical name that can be resolved through client properties.

Reply:
> The RemotingConnectorBindingInfoService.RemotingConnectorInfo does not expose that information.

It does; SocketBinding.getClientMappings(), which is used if present just above the lines shown in this diff. So this issue is only about a case where the user did not provide such information.

Arguably a remoting connector should fail if configured for the wildcard address and the SocketBinding includes no client mapping data.  But 1) that's a potentially breaking change and 2) that's not just theoretical because not all uses of a remoting connector count on the server telling the client how to connect. It's valid for the client to connect based on the user telling it how to. For example, CLI connections over the native interface work that way.

So it seems reasonable to me to do this kind of validation in a place where the intended usage is understood; i.e. where it is known that the code is meant to be telling the client how to connect.

Reply:
To me that code makes little sense. Like I said earlier, both (client & server) addresses are known when you have a connection, not at the time of binding the connector.

So after establishing a physical connection, it is easy to determine (in order of priority):

1. customer configured end point address
2. if applicable, logical connection end point address (for a clustered EJB)
3. physical connection end point address

Never would you see any `INADDR_ANY`.

(Terminology in the different man pages doesn't work out, because everything is called a bound address.)

Reply:
[The code](https://github.com/wildfly/wildfly/pull/11777/files#diff-0acb6ab0926133a24fef6c5ec5953af5R105) only gets executed when the EJB is deployed to the server, at that time, there is no client nor connection yet.

Reply:
I don't think it is easy way to update the client mappings with the physical connections at run-time. 

One possible approach may be trying to update the client mappings when the client tries to connect to server at first time. But we also need to consider how to keep them consistent, like one server in the cluster may reload so that the client mappings updated during run-time will be lost.